File: dotnet/Socrates/CandidateRegistration/CandidateProvider.cs
using System.Linq;

namespace Socrates.CandidateRegistration
{
    public class CandidateProvider : ICandidateProvider
    {
        private Candidates candidates;
        
        public CandidateProvider(params Candidate[] candidates)
        {
            this.candidates = Candidates.FromList(candidates.ToList());
        }

        public Candidates GetCandidates()
        {
            return candidates;
        }

        public void AddCandidate(Candidate candidate)
        {
            candidates.Add(candidate);
        }

        public bool IsEmpty()
        {
            return candidates.IsEmpty();
        }

        public bool HasAlready(params Candidate[] candidates)
        {
            return this.candidates.Contains(candidates);
        }

        public bool ContainsEmailsExactly(params Email[] emails)
        {
            return candidates.ContainsEmails(emails);
        }

        public Emails GetEmails()
        {
            return candidates.GetEmails();
        }
    }
}
    

File: java/src/main/java/CandidateProvider.java
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;

class CandidateProvider {

    private final Collection<Candidate> candidates = new HashSet<>();

    boolean isEmpty() {
        return candidates.isEmpty();
    }

    void add(Candidate... candidates) {
        Collections.addAll(this.candidates, candidates);
    }

    boolean exist(Candidate... candidates) {
        for (Candidate candidateCandidate : candidates) {
            if(!this.candidates.contains(candidateCandidate))
                return false;
        }
        return true;
    }

    int size() {
        return candidates.size();
    }

    Collection<Candidate> getCandidates() {
        return Collections.unmodifiableCollection(candidates);
    }
}
